core/vm/ipgraph_with_kind.go

```diff
@@ -0,0 +1,365 @@
+	totalRoyalty := big.NewInt(0)
+	if royaltyPolicyKind.Cmp(royaltyPolicyKindLAP) == 0 {
+		totalRoyalty = c.getRoyaltyLap(ipId, ancestorIpId, evm, ipGraphAddress)
+	} else if royaltyPolicyKind.Cmp(royaltyPolicyKindLRP) == 0 {
+		totalRoyalty = c.getRoyaltyLrp(ipId, ancestorIpId, evm, ipGraphAddress)
+	} else {
+		return nil, fmt.Errorf("unknown royalty policy kind")
+	}
+
+	log.Info("getRoyaltyWithKind", "ipId", ipId, "ancestorIpId", ancestorIpId, "ipGraphAddress", ipGraphAddress, "royaltyPolicyKind", royaltyPolicyKind, "totalRoyalty", totalRoyalty)
+	return common.BigToHash(totalRoyalty).Bytes(), nil
+}
+
+func (c *ipGraphWithPolicyKind) getRoyaltyStackLap(ipId common.Address, evm *EVM, ipGraphAddress common.Address) *big.Int {
+	log.Info("getRoyaltyStackLap", "ipGraphAddress", ipGraphAddress, "IP ID", ipId)
+	ancestors := make(map[common.Address]struct{})
+	totalRoyalty := big.NewInt(0)
+	var stack []common.Address
+	stack = append(stack, ipId)
+	for len(stack) > 0 {
+		node := stack[len(stack)-1]
+		stack = stack[:len(stack)-1]
+
+		currentLengthHash := evm.StateDB.GetState(ipGraphAddress, common.BytesToHash(node.Bytes()))
+		currentLength := currentLengthHash.Big()
+
+		for i := uint64(0); i < currentLength.Uint64(); i++ {
+			slot := crypto.Keccak256Hash(node.Bytes()).Big()
+			slot.Add(slot, new(big.Int).SetUint64(i))
+			storedParent := evm.StateDB.GetState(ipGraphAddress, common.BigToHash(slot))
+			parentIpId := common.BytesToAddress(storedParent.Bytes())
+
+			if _, found := ancestors[parentIpId]; !found {
+				ancestors[parentIpId] = struct{}{}
+				stack = append(stack, parentIpId)
+			}
+
+			royaltySlot := crypto.Keccak256Hash(node.Bytes(), parentIpId.Bytes(), royaltyPolicyKindLAP.Bytes()).Big()
+			royalty := evm.StateDB.GetState(ipGraphAddress, common.BigToHash(royaltySlot)).Big()
+			totalRoyalty.Add(totalRoyalty, royalty)
+		}
+	}
+	return totalRoyalty
+}
+
+func (c *ipGraphWithPolicyKind) getRoyaltyStackLrp(ipId common.Address, evm *EVM, ipGraphAddress common.Address) *big.Int {
+	log.Info("getRoyaltyStackLrp", "ipGraphAddress", ipGraphAddress, "IP ID", ipId)
+	totalRoyalty := big.NewInt(0)
+	currentLengthHash := evm.StateDB.GetState(ipGraphAddress, common.BytesToHash(ipId.Bytes()))
+	currentLength := currentLengthHash.Big()
+
+	for i := uint64(0); i < currentLength.Uint64(); i++ {
+		slot := crypto.Keccak256Hash(ipId.Bytes()).Big()
+		slot.Add(slot, new(big.Int).SetUint64(i))
+		storedParent := evm.StateDB.GetState(ipGraphAddress, common.BigToHash(slot))
+		parentIpId := common.BytesToAddress(storedParent.Bytes())
+		royaltySlot := crypto.Keccak256Hash(ipId.Bytes(), parentIpId.Bytes(), royaltyPolicyKindLRP.Bytes()).Big()
+		royalty := evm.StateDB.GetState(ipGraphAddress, common.BigToHash(royaltySlot)).Big()
+		totalRoyalty.Add(totalRoyalty, royalty)
+	}
+	return totalRoyalty
+}
+
+func (c *ipGraphWithPolicyKind) getRoyaltyLap(ipId, ancestorIpId common.Address, evm *EVM, ipGraphAddress common.Address) *big.Int {
```

Comment:
Same multiple paths related question here for `getRoyaltyLap` - are multiple paths being considered?

Would the precompile output a similar amount for `getRoyalty` and `getRoyaltyStack` as in the example below?
<img width="366" alt="image" src="https://github.com/user-attachments/assets/c6bf3f7f-a976-4bb7-ac8e-35d356fc1ff5">

Reply:
Yes, it same as lrp. all paths will be sum up.